src/OpenTelemetry.Extensions.Hosting/Implementation/EventSourceEventFormatting.cs

```diff
@@ -0,0 +1,89 @@
+// <copyright file="EventSourceEventFormatting.cs" company="OpenTelemetry Authors">
+// Copyright The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+// </copyright>
+
+using System;
+using System.Diagnostics.Tracing;
+using System.Globalization;
+using System.Linq;
+using System.Text;
+
+namespace OpenTelemetry.Internal
+{
+    internal static class EventSourceEventFormatting
+    {
+        public static string Format(EventWrittenEventArgs eventData)
+        {
+            var payloadArray = eventData.Payload?.ToArray() ?? Array.Empty<object>();
+
+            ProcessPayloadArray(payloadArray);
+
+            if (eventData.Message != null)
+            {
+                try
+                {
+                    return string.Format(CultureInfo.InvariantCulture, eventData.Message, payloadArray);
+                }
+                catch (FormatException)
+                {
+                }
+            }
+
+            var stringBuilder = new StringBuilder();
```

Comment:
Consider caching the StringBuilder + providing hint of the allocation?
https://www.dotnetperls.com/stringbuilder-cache

Reply:
Updated to use cached stringbuilder